lib/src/async.dart

```diff
@@ -0,0 +1,185 @@
+import 'dart:async';
+
+import 'package:bloc_helpers/bloc_helpers.dart';
+import 'package:meta/meta.dart';
+import 'package:rxdart/rxdart.dart';
+
+part 'request.dart';
+
+typedef Future<Result> TaskHandler<Parameter, Result>(Parameter parameter);
+
+/// Helper class to implement asynchronous task
+/// Need to implement the [run] method
+abstract class AsyncTaskBloc<Parameter, Result> extends Bloc {
+  final _callPublisher = new PublishSubject<Parameter>();
+
+  // ignore: close_sinks
+  final _resultPublisher = new PublishSubject<Result>();
+
+  // ignore: close_sinks
+  final _runningBehavior = new BehaviorSubject<bool>(seedValue: false);
+
+  AsyncTaskBloc() {
+    onCall.listen(_handleTask);
+  }
+
+  Future<void> _handleTask(Parameter input) async {
+    _runningBehavior.add(true);
+    try {
+      final res = await run(input);
+      _resultPublisher.add(res);
+    } catch (e, s) {
+      _resultPublisher.addError(e, s);
+    } finally {
+      _runningBehavior.add(false);
+    }
+  }
+
+  @mustCallSuper
+  @override
+  FutureOr<void> dispose() async {
+    await _callPublisher.close();
+
+    super.dispose();
+  }
+
+  /// Sink to start the task
+  /// the result and errors are push in the [onResult] stream
+  Sink<Parameter> get callSink => _callPublisher.sink;
+
+  /// Stream representing the current state of the bloc
+  /// true if task is ongoing
+  ValueObservable<bool> get onRunning => _runningBehavior.stream;
+
+  /// Task will start stream
+  Observable<Parameter> get onCall => _callPublisher.stream;
+
+  /// Result stream
+  Observable<Result> get onResult => _resultPublisher.stream;
+
+  /// Create a Task Bloc by passing the function
+  ///
+  /// ```dart
+  /// final taskBloc = new TaskBloc<String,int>.func(myRequest);
+  ///
+  /// Future<int> myRequest(String param) async {
+  ///   ...
+  /// }
+  /// ```
+  factory AsyncTaskBloc.func(TaskHandler<Parameter, Result> handler) =>
+      _AsyncBloc<Parameter, Result>(handler);
+
+  @protected
+  Future<Result> run(Parameter parameter);
```

Comment:
User might want to return a value directly so the return type should be `FutureOr<Result>`